spi: dw: Fix non-DMA transmit-only transfers

pelwell opened the pull request:
Ensure the transmit FIFO has emptied before ending the transfer by dropping the TX threshold to 0 when the last byte has been pushed into the FIFO. Include a similar fix for the non-IRQ paths.

See: https://github.com/raspberrypi/linux/issues/6285
Fixes: 6014649de765 ("spi: dw: Save bandwidth with the TMOD_TO feature")

pelwell committed:
spi: dw: Fix non-DMA transmit-only transfers

Ensure the transmit FIFO has emptied before ending the transfer by
dropping the TX threshold to 0 when the last byte has been pushed into
the FIFO. Include a similar fix for the non-IRQ paths.

See: https://github.com/raspberrypi/linux/issues/6285
Fixes: 6014649de765 ("spi: dw: Save bandwidth with the TMOD_TO feature")
Signed-off-by: Phil Elwell <phil@raspberrypi.com>

HiassofT commented:
It is a bit better now, but still not quite right, the TX transfer is often missing one or two clock cycles, see eg this screenshot where TX phase was one cycle short and the ID read back as 2881 instead of 5102:
![pr-2881](https://github.com/user-attachments/assets/4ba33850-c831-49e6-94d0-a635772f084f)

I've also seen one case where the kernel reported 1440 (so 2 cycles missing) and another one where the ID was read fine but a later SPI transfer failed

pelwell commented:
Although the RP1 SPI interface supports transmit-only, it lacks an interrupt-on-idle. When the TX-FIFO-empty interrupt fires, there may be some data left in the TX shift register. Spinning in interrupt context is not recommended, and by the next data is ready to be sent it may not even be necessary to wait, so we  can replace any post-TX-wait with a pre-transfer idle-wait.

If this works for you I'll squash it with the previous patch.

HiassofT commented:
Still no joy, in case of a single write-only transfer chip-select is de-asserted too early:
![pr-write-failed](https://github.com/user-attachments/assets/c77dcbd4-f7af-482c-a9d3-09f6b43d6a19)

pelwell commented:
I was hoping not to have to resort to a timer, but I can't think of an alternative with decent performance.

Third time lucky?

P33M commented on drivers/spi/spi-dw-core.c:
```diff
@@ -246,7 +272,16 @@ static irqreturn_t dw_spi_transfer_handler(struct dw_spi *dws)
 		dw_spi_mask_intr(dws, DW_SPI_INT_TXEI);
-		spi_finalize_current_transfer(dws->host);
+		if (dw_spi_ctlr_busy(dws)) {
+			dws->idle_wait_retries = 8;
```
If this is an analogue for bits per data element aka "word width", SPI can be configured up to 32. This value should be in one of the TX configuration registers...

pelwell replied:
There should be somewhere more accessible.

pelwell replied:
And there was - `dws->n_bytes`.

pelwell commented:
Apart from squashing the commits, this is the release candidate. It should reduce the number of interrupts taken to detect the end-of-transfer.

pelwell committed:
spi: dw: Clamp the minimum clock speed

The DW SPI interface has a 16-bit clock divider, where the bottom bit
of the divisor must be 0. Limit how low the clock speed can go to
prevent the clock divider from being truncated, as that could lead to
a much higher clock rate than requested.

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
